### ISSUE TITLE
# The one-shot learning for face recognition

### ISSUE THREAD

**Issue by Han-sss** [Han-sss/Hello-world#3]
# The one-shot learning for face recognition
## The seminar project - Han Sun(孫邯) Student ID: 6930-32-0777 
In this experiment, I continued the transfer learning of the previous facial recognition. Since the purpose of the experiment is to discriminate whether two pictures are the same person, the model is designed as a Siamese neural network.
Since it is transfer learning, vgg16, which was pre-trained on the object recognition training set, was used as the backbone and the Siamese network was later finetune on a homemade face dataset.

### Model design and loss convergence
In the first experiments, the model can not converge. So the last few convolutional layers of vgg16 are gradually freed in the following experiments. The results show that the model became more discriminative when a few convolutional layers are freed appropriately. But if too many convolutional layers are freed (e.g., unlocking six convolutional layers), the model will become too "aggressive" to recognize the same person. 
However, in subsequent quantitative experiments, it is shown that the model is still valid even if the loss does not converge significantly.

### Margin of the Contrastive  loss function
In past experiments, the effect of the value of margin in the Contrastive loss function on the model is also explored. Margin is similar to a magnification factor that makes the model performs more "aggressive". When the output is less than half of the maximum value, a large margin value will make the output smaller. Conversely, it makes the output larger. This means that when the model thinks that two people are not very similar, the great margin will enlarge the difference between them, in contrast, it will reduce the difference between them.
![image](https://user-images.githubusercontent.com/63163321/124904329-d9589000-e01f-11eb-8cf3-0cee30638058.png)
The picture of the second column is the result of the model with a great margin value.

### My own dataset
In addition, I built my own dataset with images sampled from videos on the web, mainly from news, talk shows, and interviews of all kinds. I did this because the previous dataset is gray-scale, and the pre-trained model was trained on an RGB dataset. In addition, the orientations of the faces in the original dataset varied too much, so I removed the background and limited the orientations of the faces. Finally, I further cropped the dataset to ensure that only the facial features were preserved.
![image](https://user-images.githubusercontent.com/63163321/124902165-a7dec500-e01d-11eb-9999-21576f31819c.png )
![image](https://user-images.githubusercontent.com/63163321/124902190-af05d300-e01d-11eb-8910-3825de4337f2.png)
The first row of figures is from my own dataset. The second row of figures is cropped.

### Simulation experiment
Finally, I also inserted simulation experiments in this report. The two input images are from the training and validation datasets of the same person. This is used to simulate the recognition of the images captured from the camera. This experiment was performed on the cropped dataset. The results are shown below.

CONV4: False False False; CONV5: False, False, False
![cat (6)](https://user-images.githubusercontent.com/63163321/124912863-6bb16180-e029-11eb-9d81-0c88f09023b6.png)
CONV4: False False False; CONV5: False, False, True
![cat](https://user-images.githubusercontent.com/63163321/124904677-381e0980-e020-11eb-854b-92238a63a916.png)
CONV4: False False False; CONV5: False, True, True
![cat (1)](https://user-images.githubusercontent.com/63163321/124904700-3ce2bd80-e020-11eb-9c5c-df70b9ac76a3.png)
CONV4: False False False; CONV5: True, True, True
![cat (2)](https://user-images.githubusercontent.com/63163321/124904745-45d38f00-e020-11eb-9446-09f3bc7e5ece.png)
CONV4: False False True; CONV5: True, True, True
![cat (3)](https://user-images.githubusercontent.com/63163321/124905918-7ff16080-e021-11eb-87d2-1840925fdd4a.png)
CONV4: False TrueTrue; CONV5: True, True, True
![cat (4)](https://user-images.githubusercontent.com/63163321/124907453-238f4080-e023-11eb-8edb-3e481005718e.png)
CONV4: True TrueTrue; CONV5: True, True, True
![cat (5)](https://user-images.githubusercontent.com/63163321/124908578-6c93c480-e024-11eb-887f-cd079bb79148.png)

After freeing three convolutional layers, the model starts to become overfitted. All the result values of the difference between the images in the training set and the images in the validation set are gradually increasing. But in every group, the model still tends to think they are the same person.

About the female in the third group, the result values are far greater than expected due to the poor freedom of the model. The problem is mitigated after more convolutional layers are freed later.
